### DISCUSSION TITLE
Building Multi-Tenent app: Domain added via Domain API returns status code 307

### DISCUSSION THREAD

**Discussion by chavda-bhavik** [vercel/community#3100] [Help] [answered]
### Summary

I'm building a multi-tenant app where I'm adding a domain via [domain api](https://vercel.com/docs/rest-api/endpoints#add-a-domain-to-a-project), but even after the user verifies the domain, the browser returns 307 status code always.

Here is the CURL,
```
curl --location 'https://api.vercel.com/v10/projects/abcd/domains?teamId=abcd' \
--header 'Authorization: Bearer XXXX' \
--header 'Content-Type: application/json' \
--data '{
    "name": "www.bhavikchavda.com",
    "redirect": "bhavikchavda.com"
}'
```
Here is the screenshot of the vercel logs,
![image](https://github.com/vercel/community/assets/50201755/80b08d59-5d8d-481e-a57b-bdf4d015d5e9)

I found out that the issue is with redirection, here domain is showing no-redirect with status code 307,
![image](https://github.com/vercel/community/assets/50201755/e4794ce8-2f39-4827-a8d8-38019c7e7970)

Here is the response of [verify-api](https://vercel.com/docs/rest-api/endpoints#verify-project-domain),
```
{
    "name": "www.bhavikchavda.com",
    "apexName": "bhavikchavda.com",
    "projectId": "XXXX",
    "redirect": "bhavikchavda.com",
    "redirectStatusCode": null,
    "gitBranch": null,
    "updatedAt": 1688350035132,
    "createdAt": 1688349983847,
    "verified": true
}
```

When adding the domain from the vercel dashboard it shows two entries, one is for `bhavikchavda.com` and another is for `www.bhavikchavda.com`, but while adding domain via domains-api only shows one entry,
![image](https://github.com/vercel/community/assets/50201755/7aef773d-a516-4cc6-a4d0-c2c01186d020)

### Steps to Reproduce

1. Add the domain via [Domain API](https://vercel.com/docs/rest-api/endpoints#add-a-domain-to-a-project).
2. Add domain `CNAME` and `A` record entries to the domain provider.
3. Once the domain is verified, see the domain entry in vercel, it will show the domain entry like this, and hitting the domain in the browser will show the error `domain redirected you many times` because of the 307 status code.
![image](https://github.com/vercel/community/assets/50201755/e4794ce8-2f39-4827-a8d8-38019c7e7970)

**Comment by amyegan** [comment]
Would you be able to provide an example of what you're sending in the api call? Censor any sensitive information, of course, but I want to make sure I'm testing with the same data shape

**Reply by chavda-bhavik** [reply]
Of course @amyegan, Here is the CURL,
```
curl --location 'https://api.vercel.com/v10/projects/XXXX/domains?teamId=XXXX' \
--header 'Authorization: Bearer XXXX' \
--header 'Content-Type: application/json' \
--data '{
    "name": "www.bhavikchavda.com",
    "redirect": "bhavikchavda.com"
}'
```
- `XXXX` are censored values.
- After adding the domain, you can add `CNAME` or `TXT` values in the domain provider and can see that only one entry is shown in the vercel dashboard with the `307` status code.

Thanks for looking in.

**Reply by smaeda-ks** [reply]
I'll take a look

**Reply by smaeda-ks** [reply]
ok, I've just submitted a PR internally to mitigate this. we will probably disallow such an operation going forward (i.e., adding domains with `redirect` set to a domain that isn't added to the project) to prevent this. Thanks for flagging this @chavda-bhavik

**Reply by chavda-bhavik** [reply]
@smaeda-ks If I understood you properly, you're saying adding a custom domain that is not already added to the vercel dashboard will be removed.

But it will lead problem to many users like us. Our use-case is we're building a multi-tenant app where users can add their custom domain name like `www.example.com`. And we're providing instructions to users for configuring the `CNAME` or `TXT` records in their domain providers.

Do you mean to disallow this kind of operation?

**Reply by chavda-bhavik** [reply]
Or do we have to call [add-domain](https://vercel.com/docs/rest-api/endpoints#add-a-domain-to-a-project) API two times?

**Reply by smaeda-ks** [reply, marked as answer]
> Or do we have to call add-domain API two times?

That is correct. You will need to first add `havikchavda.com`, then add `www.havikchavda.com` with redirect set to `havikchavda.com`. That's what we're doing on the UI as well.

**Reply by chavda-bhavik** [reply]
Thanks @smaeda-ks for the information. Can I implement this flow in the application to get it working?

**Reply by smaeda-ks** [reply]
yes, you can implement it that way if you're always adding that www to Apex redirect for new onboarding. The fix I'm trying to add is just to ensure we return an error if you add a domain with a redirect set to a domain that's not a part of the project.

**Reply by smaeda-ks** [reply]
Alright, the API change is now live. You should see 400 error if you attempt the original reproduction steps above.
